pkg/providers/vault/client.go

```diff
@@ -27,32 +29,136 @@ import (
+		// token is not renewable, log warning and return
+		c.lc.Warn(fmt.Sprintf("token '%s' is not renewable from the secret store", token))
+		return nil
+	}
+
+	// the renew interval is half of period value
+	tokenPeriod := time.Duration(tokenData.Data.Period) * time.Second
+	renewInterval := tokenPeriod / 2
+	if renewInterval <= 0 {
+		// no renew
+		c.lc.Warn(fmt.Sprintf("token '%s' renewInterval is 0, no token renewal", token))
+		return nil
+	}
+
+	ttl := time.Duration(tokenData.Data.Ttl) * time.Second
+
+	// if the current time-to-live is already less than the half of period
+	// need to renew the token right away
+	if ttl <= renewInterval {
+		// call renew self api
+		c.lc.Info("ttl already <= half of the renewal period")
+		if err := c.renewToken(); err != nil {
+			return err
 		}
-		config.retryWaitPeriodTime = retryTimeDuration
 	}
 
-	return Client{
-		HttpConfig: config,
-		HttpCaller: httpClient,
-	}, nil
+	// goroutine to periodically renew the service token based on renewInterval
+	go func() {
+		ticker := time.NewTicker(renewInterval)
+		for {
+			select {
+			case <-ctx.Done():
+				ticker.Stop()
+				c.lc.Info("context cancelled, dismiss the token renewal process")
+				return
+			case <-ticker.C:
+				// renew token
+				// if err happens then stop the ticker and return
+				if err := c.renewToken(); err != nil {
+					errChan <- err
+					ticker.Stop()
+					return
+				}
+			}
+		}
+	}()
+
+	return nil
+}
+
+func (c Client) getTokenLookupResponseData() (*TokenLookupResponse, error) {
+	// call Vault's token self lookup API
+	url := c.HttpConfig.BuildURL() + "/v1/auth/token/lookup-self"
+	req, err := http.NewRequest(http.MethodGet, url, nil)
+	if err != nil {
+		return nil, err
+	}
 
+	req.Header.Set(AuthTypeHeader, c.HttpConfig.Authentication.AuthToken)
+
+	resp, err := c.HttpCaller.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, pkg.NewErrSecretStore(fmt.Sprintf("tried to lookup token but received a '%d' response from the secret store", resp.StatusCode))
+	}
+
+	defer resp.Body.Close()
+	var result TokenLookupResponse
+	jsonDec := json.NewDecoder(resp.Body)
+	if jsonDec == nil {
+		return nil, pkg.NewErrSecretStore("failed to obtain json decoder")
+	}
+
+	jsonDec.UseNumber()
+	if err = jsonDec.Decode(&result); err != nil {
+		return nil, err
+	}
+
+	return &result, nil
+}
+
+func (c Client) renewToken() error {
+	// call Vault's renew self API
+	url := c.HttpConfig.BuildURL() + "/v1/auth/token/renew-self"
+	req, err := http.NewRequest(http.MethodPost, url, nil)
+	if err != nil {
+		return err
+	}
+
+	req.Header.Set(AuthTypeHeader, c.HttpConfig.Authentication.AuthToken)
+
+	resp, err := c.HttpCaller.Do(req)
+	if err != nil {
+		return err
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return pkg.NewErrSecretStore(fmt.Sprintf("tried to renew token but received a '%d' response from the secret store", resp.StatusCode))
+	}
+
+	defer resp.Body.Close()
```

Comment:
Should the defer be before line 155 (if error return)?

Reply:
good catch